### ISSUE TITLE
Playing Doubles charts in Solo?

### ISSUE THREAD

**Issue by TonyTheTGR** [stepmania/stepmania#682]
Was thinking of a script that takes the doubles column, and just overlaps a couple of the layers like:

[3] [1] [4] [2] [5] [3] [6] [4]

...so that if Doubles charts are present, those would generate more cohesive 6-panel DDR charts. No idea where to begin that, though.

"Feature Req"

**Comment by nixtrix** [comment]
Could be possible, but why would you want that? I'm not sure it'd make for very playable results at all. Have you tested it out?

**Comment by kyzentun** [comment]
The current logic for picking the chart to autogen from looks for the steps type with the most similar number of tracks.  So if it needs to autogen a chart with 6 tracks, it prefers 5 or 7 tracks over 4 or 8 tracks, and 4 or 8 tracks over 3 or 9 and so on.
https://github.com/stepmania/stepmania/blob/master/src/Song.cpp#L1368-L1384

The only reason the current logic selects the 4 track chart to generate from instead of the 8 track chart is that StepsType_DanceSingle comes before StepsType_DanceDouble.

The actual generation algorithms are in NoteDataUtil, with LoadOverlapped handling the case where the source note data has more tracks and LoadTransformedSlidingWindow handling the case where the source has fewer tracks.
LoadOverlapped has some logic to keep two source tracks from putting notes into the same destination track too close to each other, so it's better to just use the already existing autogen logic than to write something that overlaps tracks without considering that problem.

As a side note, neither autogen method actually generates charts that are enjoyable on pad.  If that's your goal, you'd have to start from scratch and base it on a serious and in depth consideration of the mechanics of motion during play.

**Comment by TonyTheTGR** [comment]
You'd be surprised at how well it flows. I'd manual-edited a few charts like that to test it, and it's not perfect, but no autogenerator will be. That said, double pivots, quite literally, on it's inner-two side arrows, just as Solo does on it's diags... and then the rest is just keeping the down/outside/up consistent. So, it's kind of like just playing the real double chart, and using the diags as the center arrows.

And... what I can just jump in and start mudding up code? O.o I'm fresh enough to StepMania source to know better than to do that to you all! ^.^'

**Comment by kyzentun** [comment]
People can make pull requests that implement what they want, and if the main devs dislike the pull request, there's usually an explanation why.

**Comment by nixtrix** [comment]
Have you used github to add stuff to other people's projects before?

**Comment by TonyTheTGR** [comment]
Nope. Just signed up a few days ago. Between SM developments, and having a
HDD tank on me while coding a project, it was recommended I begin one.
Still new to Git, though. Gittin' the feel of things, still.
